Creates PoliticalMap.kt

ericbrandwein opened the pull request:
Represents the map with its countries, continents, and borders.

ericbrandwein committed:
Creates PoliticalMap.kt

Represents the map with its countries, continents,
and borders.

nsanmartin commented on core/src/main/PoliticalMap.kt:
```diff
@@ -0,0 +1,72 @@
+class PoliticalMap private constructor(
+    private val countriesWithContinents: Map<String, String>,
+    private val borders: Map<String, Collection<String>>
+) {
```
No se si representar paises y continentes directamente on strings. Preferiría usar o typedes (tipo c) si hay, o (mejor) hacer clases para cada cosa, tipo:

`class Country constructor(val name : String)`

Eso esta bueno porque ya tenemos una clase y eventualmente si lo necesitaoms podemos agregar atributos sin que el refactor sea demasiago grande (no haría que cambiar la signatura de todas las funciones que usan Country). Lo mismo para continent.

ericbrandwein replied:
Yo pense lo mismo pero después pensé que si el country es solo un String no servía agregar complejidad. Aunque creo que hay typedefs, así que se podría hacer eso.

Lo otro que se podría hacer es que el politicalMap tenga los countries y continents como tipos paramétricos, y ahí si querés ponerle una clase después es fácil, es sólo cambiar donde lo inicializás (y donde se usa, pero bueno). 

Hay tres opciones: [typedef](https://www.google.com/search?client=firefox-b-d&q=kotlin+typedef), clase Country, o tipo paramétrico. Cuál te gusta más?

nsanmartin replied:
Gustarme me gusta mas tipos param'etricos ;), pero no s'e si vale la pena en este caso porque no ve oque haya como un concepto de pa'is y paises de muchos tipos , etc. Si typedef funciona e nrealidad como 'using' de c++ (o sea se toma un tipo nuevo y no compila con el tipo del que es alias) entonces podriamos arrancar con typedesf hasta que sea necesario un refactor si queremos que tenga algun otro comportamiento o dato distinto de string. Que sea param'etrico el politicalMap es independiente, porque puede serlo aunque hicieramos clases para countries . Yo enotnces  esto lo dejaria asi y is queres hacemo suna task para reflexionar sobre si queremo so no tipos parametricoa ac'a.

ericbrandwein replied:
el typealias no hace nada más que hacer que sea más fácil escribir, no cambia nada en compilación.

nsanmartin replied:
class Country entonces

nsanmartin commented on core/src/main/PoliticalMap.kt:
```diff
@@ -0,0 +1,72 @@
+class PoliticalMap private constructor(
+    private val countriesWithContinents: Map<String, String>,
+    private val borders: Map<String, Collection<String>>
+) {
```
No estoy seguro de estar de acuerdo de representar el grafo con un map. Acá no hay un tema de eficiencia algorítmica ni nada (la cantidad de países está acotada por una constante suficientemente grande, salvo que hagamos que el PolticalMap sea un bosque, con muuchos árboles y que cada arbol tenga sus objetivos y después se pase a objetivos a gra nescala y con eso y otras cosas asi pueden jugar muchísimos jugadores...). Se me hace más simple  pensar en los continentes como listas (ordenadas para hacer búsquedas binarias) . Las adyascencias sí me gusta como map de pais a lista de pais. ¿Existe el tipo SortedList acá? Sería preferible porque el mapa lo hacemons una vez al principio y lo consultamos todo el tiempo

ericbrandwein replied:
Me parece bien. El SortedList puede ser un Set, que ya te viene con búsqueda en log(n). Depende mucho de qué vas a estar usando más: si vas a querer ver el continente de un país o los países de un continente. Creo que vas a querer preguntar más los países de un continente, porque para saber si conquistaste todo el continente tenés que preguntar cuáles son sus países, así que la representación de listas aparte de ser más clara es mejor. Ahora lo cambio.

ericbrandwein replied:
Ahí lo cambié, e hice dos commits sobre esto: el primero usa el map con los sets, el segundo agrega una clase Continent que agrupa el nombre y los countries en una sola cosa. Lo hice para que sea más claro qué estamos haciendo con un mapa de string a collection. Quizá no hace falta, especialmente con los typealias que podemos poner para Continent y Country, pero no sé, fijate cuál te gusta más.

nsanmartin commented on core/src/main/PoliticalMap.kt:
```diff
@@ -0,0 +1,72 @@
+        return borders.getValue(firstCountry).contains(secondCountry)
+    }
+
+    class Builder {
```
¿Porqué un "Builder", no alcanza con hacer un constructor de PoliticalMap y métodos de políticalMap? (es pregunta) Es decir, me imagino un constructor que recibe los continentes y las adyascencias y checkea lo que tenga que ckequear.

ericbrandwein replied:
Porque quería que el PoliticalMap sea inmutable pero quería también poder crearlo con métodos lindos y claros, no pasándole un mapa y una lista así tirados.

nsanmartin replied:
Medio al márgen y volado: no podria ser divertido que los paises cambien? onda se puede dividir un pais ponele uruguay del norte uruguay del sur...

Lo del builder lo disutimos en el canal de filosofía

ericbrandwein replied:
ahh tipo en el medio del juego? Tendríamos que agregar reglas especiales para saber cuándo se separan dos países. Uh imaginate que tanto como que se pueden separar países se pudiesen unir, y lo malo de unirlos es que hay más probabilidades de que te aparezcan grupos internos que quieran guerra civil, y lo bueno de unirlos es que en vez de tener muchos países con un solo ejército tenés uno solo con muchos ejércitos y por lo tanto puede atacar. Igual ya nos fuimos por las ramas, y sería otro juego completamente diferente.

ericbrandwein committed:
Map of continent to countries in PoliticalMap

ericbrandwein committed:
Creates Continent class in PoliticalMap

ericbrandwein committed:
Type alias Country = String in PoliticalMap

It doesn't change anything in compilation, it's just
to make it easier to understand.
